examples/middleware/examples/builder.rs

```diff
@@ -0,0 +1,56 @@
+use ethers_core::types::BlockNumber;
+use ethers_middleware::{gas_escalator::*, gas_oracle::*, *};
+use ethers_providers::{Http, Middleware, Provider};
+use ethers_signers::{LocalWallet, Signer};
+use std::convert::TryFrom;
+
+const RPC_URL: &str = "https://mainnet.infura.io/v3/c60b0bb42f8a4c6481ecd229eddaca27";
+const SIGNING_KEY: &str = "fdb33e2105f08abe41a8ee3b758726a31abdd57b7a443f470f23efce853af169";
+
+/// In ethers-rs, middleware is a way to customize the behavior of certain aspects of the library by
+/// injecting custom logic into the process of sending transactions and interacting with contracts
+/// on the Ethereum blockchain. The MiddlewareBuilder trait provides a way to define a chain of
+/// middleware that will be called at different points in this process, allowing you to customize
+/// the behavior of the Provider based on your needs.
+#[tokio::main]
+async fn main() {
```

Comment:
We should add an example of writing your own middleware. And an example showing how to pass middlewares around as M: Middleware. I've seen people get really confused at the type errors because they try to pass eg SignerMiddleWare<LocalWallet, Provider<Http>>

Reply:
Hi @gakonst
I provided an example in [custom_gas_middleware.rs](https://github.com/gakonst/ethers-rs/pull/1940/files#diff-6be7519191da8a934a30d0c69f6424106a3c1af298113f5db73d23d329af6575)

It provides a constructor that takes `inner: M` as an argument with a trait bound on M. 
If you remember specific cases where users got confused please let me know.